service/controller/v13/resource/instance/create_cordon_old_workers.go

```diff
@@ -0,0 +1,161 @@
+package instance
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-06-01/compute"
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/errors/tenant"
+	"github.com/giantswarm/microerror"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/giantswarm/azure-operator/service/controller/v13/controllercontext"
+	"github.com/giantswarm/azure-operator/service/controller/v13/key"
+	"github.com/giantswarm/azure-operator/service/controller/v13/resource/instance/internal/state"
+)
+
+const (
+	// UnschedulablePatch is the JSON patch structure being applied to nodes using
+	// a strategic merge patch in order to cordon them.
+	UnschedulablePatch = `{"spec":{"unschedulable":true}}`
+)
+
+func (r *Resource) cordonOldWorkersTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	if cc.Client.TenantCluster.K8s == nil {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "tenant cluster client not available yet")
+		return currentState, nil
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding all worker VMSS instances")
+
+	var allWorkerInstances []compute.VirtualMachineScaleSetVM
+	{
+		allWorkerInstances, err = r.allInstances(ctx, customObject, key.WorkerVMSSName)
+		if IsScaleSetNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find the scale set '%s'", key.WorkerVMSSName(customObject)))
+
+			return currentState, nil
+		} else if err != nil {
+			return "", microerror.Mask(err)
+		}
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d worker VMSS instances", len(allWorkerInstances)))
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding all tenant cluster nodes")
+
+	var nodes []corev1.Node
+	{
+		nodeList, err := cc.Client.TenantCluster.K8s.CoreV1().Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+		nodes = nodeList.Items
+	}
+
+	oldNodes, newNodes := sortNodesByTenantVMState(nodes, allWorkerInstances, customObject, key.WorkerInstanceName)
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d old and %d new nodes from tenant cluster", len(oldNodes), len(newNodes)))
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring old nodes are cordoned")
+
+	oldNodesCordoned, err := r.ensureNodesCordoned(ctx, oldNodes)
+	if tenant.IsAPINotAvailable(err) {
```

Comment:
How come we need to check if the API is available here?

Reply:
Good point. Might just as well return wrapped error. I'll change.